accepted/future-releases/variance/feature-specification.md

```diff
@@ -0,0 +1,418 @@
+# Sound and Explicit Variance
+
+Author: eernst@google.com
+
+Status: Draft
+
+
+## CHANGELOG
+
+2019.08.30:
+- Initial version uploaded.
+
+
+## Summary
+
+This document proposes a specification for sound and explicit management of [variance](https://github.com/dart-lang/language/issues/213) in Dart.
+
+Issues on topics related to this proposal can be found [here](https://github.com/dart-lang/language/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3Avariance+).
+
+Currently, a parameterized class type is covariant in every type parameter. For example, `List<int>` is a subtype of `List<num>` because `int` is a subtype of `num` (so the list type and its type argument "co-vary").
+
+This is sound for all covariant occurrences of such type parameters in the class body (for instance, the getter `first` of a list has return type `E`, which is sound). It is also sound for contravariant occurrences when a sufficiently exact receiver type is known (e.g., for a literal like `<num>[].add(4.2)`, or for a generative constructor `SomeClass<num>.foo(4.2)`).
+
+However, in general, every member access where a covariant type parameter occurs in a non-covariant position may cause a dynamic type error, because the actual type annotation at run time&mdash;say, the type of a parameter of a method&mdash;is a subtype of the one which occurs in the static type.
+
+This proposal introduces explicit variance modifiers for type parameters, as well as invariance modifiers for actual type arguments. It includes compile-time restrictions on type declarations and on the use of objects whose static type includes these modifiers, ensuring that the above-mentioned dynamic type errors cannot occur.
+
+In order to ease the transition where types with explicit variance are created and used, this proposal allows for certain subtype relationships where dynamic type checks are still needed when using legacy types (where type parameters are _implicitly_ covariant) to access an object, even in the case where the object has a type with explicit variance. For example, it is allowed to declare `class MyList<out E> implements List<E> {...}`, even though this means that `MyList` has members such as `add` that require dynamic checks and may incur a dynamic type error.
+
+
+## Syntax
+
+The grammar is adjusted as follows:
+
+```
+<typeParameter> ::= // Modified.
+    <metadata> <typeParameterVariance>? <typeIdentifier>
+    (('extends'|'super') <typeNotVoid>)?
+
+<typeParameterVariance> ::= // New.
+    'out' | 'inout' | 'in'
```

Comment:
Since out and inout have no special status, presumably it's valid to use them as type parameter names?  And even with variance modifiers?  e.g. `class C<out out> {}` is a valid declaration of a class with one covariant type parameter named `out`?

Reply:
cc @kallentu